Chess, Sofas & Shutters - Furniture Update

KathrinBailey opened the pull request:
## About The Pull Request
A series of changes bulked together to give players and mappers a little more freedom in their builds to make things look nicer. 

**Chess:**
• Introduces a bunch of jumbo chess pieces that are craftable in-game. Sprites are my own.

**Shutters**
• Defines 100 integrity for the shutters and gives them an armor block so they don't inherit their stats from blast doors. This means shutters will no longer be just a reskinned blast door.
• Redid the sprites for the shutters. They now animate much smoother and look like normal slanted shutters. Their blinking light once animating is also easier to see.
• Added the window shutter; this is a normal shutter with a window so you can always see past it. I've used it in prisons and on armouries in the codebase I'm from to allow officers to see into the areas shutters are protecting, giving the area additional protection of having someone being able to look into it.
• Added radiation shutters. Maintainers feel free to question this. Through extensive experimentation, any rad insulation value between 0 and 1 does pretty much nothing to radiation. Numbers higher than 1 start multiplying it, negative numbers start reducing it. This will stop the radiation from a supermatter irradiating you **only if you stand behind it**. If you're not covered by the shutters or the shutters are open, they won't protect you at all. They also have their own sprite with a radioactive hazard symbol.
• Added the rad shutters to in-rotation maps on their supermatter engines and the windowed shutters to Metastation armoury.

**Sofas**
• Introduces the corporate sofa. A sleek grey sofa. The original sprite I ported from Hyperstation's files. I'm not sure where they got the sofa but I was told it was Eris. I didn't see it there but I didn't look super hard. I've updated the sprites on it to work with armrest sprite overlays.
• Ports the sofa sprites from Cit and the armrest sprite overlays. They look nicer than awkwardly floating over the sofa. The sofas are the same, they're just moved a little for the armrest to work.
• Makes the new corporate sofa craftable.

## Why It's Good For The Game
**Why are the chess pieces good for the game?**
I'm aware there's some sort of chess tablet already, but it's a tad impersonal. Jumbo chess pieces allow you to make a spectacle out of your chess match. Who's rolling up to the station with the nuke? The Syndicate... Chess Grandmaster? Who wants to challenge the Captain to a game in front of the crew? Sure if you want to do something silly like that, you can. I personally played chess on the start screen for fun, OOC chat seemed to enjoy it. **See below for images of the sprites.**

**Why are the sofa changes good for the game?**
Floating awkwardly over the sofas doesn't look too great. The armrest overlays make it look like you're sitting. I of course take no credit for that, the only part of the sofa changes that isn't a port is the (minor) updates to the ported corporate sofa sprite so they work with armrests. I originally ported this to Cit, and Cit had armrests, so...
For mappers, this new sofa allows you to give areas a more modern aesthetic. In Nanotrasen's state-of-the-art Space Station 13, having a single sofa that isn't "old" and "ratty" is cool..

**Why are the shutter changes good for the game?**
The new sprites were received well when I did them on other servers. The animation is much smoother.
Now, radiation shutters. It doesn't make sense we have shutters that don't actually do anything just outside the SM chamber. These radiation shutters don't really offer much protection unless you hug them. They're mostly aesthetic, but they do help just a tiny bit. Of course, once there is a delamination or a very radioactive build the shutters won't help as much. It also looks nice to have the radioactive hazard symbol on the shutters right outside the chamber. Again, if you aren't directly behind the shutters you're getting irradiated just as much as if they weren't there at all.

## Changelog
:cl: Floof Ball / Kathrin Morrison
add: Craftable jumbo chess pieces. Have an 8x8 space on the station? Have a public match people can watch!
add: New shutter sprites. Adds windowed and radiation-proofed shutters.
add: New sofa, craftable in-game with metal like the regular sofas.
tweak: Sofas now have working armrests.
/:cl:

## Acknowledgements
Citcode for the armrest overlays.
Hyperstation for the corporate sofa's original sprites, although they may have taken it from CEV Eris.
Rohesie for being a kind point of contact for TG.

## Media
![image](https://user-images.githubusercontent.com/53862927/93706581-9294ba80-fb1f-11ea-999f-a5ab6139eeff.png)
_Chess pieces!_ **Note: This isn't on a TG map.**

![image](https://user-images.githubusercontent.com/53862927/93706594-adffc580-fb1f-11ea-8e19-774026fbe474.png)
_Shutter sprites._

![image](https://user-images.githubusercontent.com/53862927/93706619-ddaecd80-fb1f-11ea-9e67-9eb6b2218c91.png)
_Windowed shutters closing off a cell block in prison. If there was a riot, rioters can be contained and officers can see what they're facing once they open the shutters to retake the wing._ **Note: This isn't on a TG map.**

![image](https://user-images.githubusercontent.com/53862927/93706628-fa4b0580-fb1f-11ea-99e4-64fbed7aabe1.png)
_One of the intended uses of windowed shutters for mappers, to close the armoury off whilst allowing the Warden and other Officers to still see inside of it._ **Note: This isn't on a TG map.**

![image](https://user-images.githubusercontent.com/53862927/93706685-84936980-fb20-11ea-9aa2-8adc23ae80e8.png)
_Me playing chess on the start screen to entertain people._ **Note: This isn't on TG. Although I think that's obvious.**

![image](https://user-images.githubusercontent.com/53862927/93706732-35016d80-fb21-11ea-949a-39e671e1e8af.png)
_How the armrest overlays look on the new sofa in Metastation bar._

![image](https://user-images.githubusercontent.com/53862927/93706879-77777a00-fb22-11ea-8ceb-4ebbdfa42d22.png)
_How the armrest overlays look on the original sofa in the middle of nowhere._

![image](https://user-images.githubusercontent.com/53862927/93706741-55312c80-fb21-11ea-825d-70bf54279ba9.png)
_An example of how the new sofas can be used in the map editor._

## Testing
I've tested this locally, the shutters all work as intended. Radiation shutters included. All the armrest overlays work for all armrests and corner sofa bits, you can craft and deconstruct the chess pieces, make and rotate the sofas still.
_As silly as this bit sounds I'm just trying to save maintainers a bit of time. The only thing that maintainers should have to be concerned about with this is art direction and if negative rad insulation causes issues. In regards to that, they've been merged since May 26th on Citadel and May 29th on Skyrat and on both servers, it hasn't caused any issues._

KathrinBailey committed:
Shutters

Adds new shutter sprites and window + radiation shutters. Windows shutters are always see-through and radiation shutters block low to moderate amounts of radiation.

KathrinBailey committed:
Chess, Map Update & quick shutter fix

Accidently pressed up arrow key on one of the new shutters, making the top two rows of pixels go to the bottom. Oops, fixed.

Radiation shutters are on the map. On advice from Rohesie, so there's no need for a follow-up PR.

Metastation has windowed shutters on the armoury instead of normal ones.

Chess pieces are introduced here. Check PR for pic.

Corporate Sofas are here too. Originally I ported this from Hyperstation to Citadel and updated sprites so it works with buckling. I think Hyper ported this from Eris? But I've never seen it.

KathrinBailey committed:
Sofa armrest overlays ported

Apparently sofa armests didn't exist in TG. This is a quick port, because I assumed it was just here already. Sprites for the sofas where the armrest works have been ported also.

Rohesie commented on code/game/machinery/doors/shutters.dm:
```diff
@@ -15,3 +17,36 @@
+	rad_insulation = 1
+
+/obj/machinery/door/poddoor/shutters/radiation/do_animate(animation)
+	..()
```
```suggestion
	. = ..()
```
Unless you have an explicit reason not to, preserve the parent call's return.

KathrinBailey committed:
Rohesie's suggested change.

Co-authored-by: Rohesie <rohesie@gmail.com>

Rohesie commented on code/game/machinery/doors/shutters.dm:
```diff
@@ -15,3 +17,36 @@
+	desc = "A shutter with a thick see-through polycarbonate window."
+	icon = 'icons/obj/doors/shutters_window.dmi'
+	icon_state = "closed"
+	opacity = 0
+	glass = 1
```
```suggestion
	opacity = FALSE
	glass = TRUE
```
* Defines for booleans, always.

KathrinBailey committed:
Rohesie's suggested change.

Co-authored-by: Rohesie <rohesie@gmail.com>

Rohesie commented on code/game/machinery/doors/shutters.dm:
```diff
@@ -15,3 +17,36 @@
+	opacity = 0
+	rad_insulation = 1
+
+/obj/machinery/door/poddoor/shutters/radiation/do_animate(animation)
+	. = ..()
+	switch(animation)
+		if("opening")
+			rad_insulation = 1
+		if("closing")
+			rad_insulation = -0.5	//Rad insulation 0 to 1 doesn't really do anything to radiation, above 1 starts multiplying it, you need to go into negative numbers to start blocking radiation.
```
This code needs to go into the opening and closing procs, instead of the animation one.

KathrinBailey replied:
It now does this on open() and close().
Tested and working as intended.
![image](https://user-images.githubusercontent.com/53862927/93707732-0d160800-fb29-11ea-99a6-9f5c853d7f46.png)
![image](https://user-images.githubusercontent.com/53862927/93707736-18693380-fb29-11ea-96ac-8c31c595a458.png)

Rohesie commented on code/game/objects/items/stacks/sheets/sheet_types.dm:
```diff
@@ -35,7 +35,27 @@ GLOBAL_LIST_INIT(metal_recipes, list ( \
+		new /datum/stack_recipe("sofa (corner)", /obj/structure/chair/sofa/corp/corner, one_per_turf = TRUE, on_floor = TRUE), \
+		)), \
+	new /datum/stack_recipe_list("chess pieces", list( \
+		new /datum/stack_recipe("White Pawn", /obj/structure/chess/whitepawn, 2, time = 10, one_per_turf = 1, on_floor = 1), \
+		new /datum/stack_recipe("White Rook", /obj/structure/chess/whiterook, 2, time = 10, one_per_turf = 1, on_floor = 1), \
+		new /datum/stack_recipe("White Knight", /obj/structure/chess/whiteknight, 2, time = 10, one_per_turf = 1, on_floor = 1), \
+		new /datum/stack_recipe("White Bishop", /obj/structure/chess/whitebishop, 2, time = 10, one_per_turf = 1, on_floor = 1), \
+		new /datum/stack_recipe("White Queen", /obj/structure/chess/whitequeen, 2, time = 10, one_per_turf = 1, on_floor = 1), \
+		new /datum/stack_recipe("White King", /obj/structure/chess/whiteking, 2, time = 10, one_per_turf = 1, on_floor = 1), \
+		new /datum/stack_recipe("Black Pawn", /obj/structure/chess/blackpawn, 2, time = 10, one_per_turf = 1, on_floor = 1), \
+		new /datum/stack_recipe("Black Rook", /obj/structure/chess/blackrook, 2, time = 10, one_per_turf = 1, on_floor = 1), \
+		new /datum/stack_recipe("Black Knight", /obj/structure/chess/blackknight, 2, time = 10, one_per_turf = 1, on_floor = 1), \
+		new /datum/stack_recipe("Black Bishop", /obj/structure/chess/blackbishop, 2, time = 10, one_per_turf = 1, on_floor = 1), \
+		new /datum/stack_recipe("Black Queen", /obj/structure/chess/blackqueen, 2, time = 10, one_per_turf = 1, on_floor = 1), \
+		new /datum/stack_recipe("Black King", /obj/structure/chess/blackking, 2, time = 10, one_per_turf = 1, on_floor = 1), \
```
```suggestion
		new /datum/stack_recipe("White Pawn", /obj/structure/chess/whitepawn, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
		new /datum/stack_recipe("White Rook", /obj/structure/chess/whiterook, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
		new /datum/stack_recipe("White Knight", /obj/structure/chess/whiteknight, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
		new /datum/stack_recipe("White Bishop", /obj/structure/chess/whitebishop, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
		new /datum/stack_recipe("White Queen", /obj/structure/chess/whitequeen, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
		new /datum/stack_recipe("White King", /obj/structure/chess/whiteking, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
		new /datum/stack_recipe("Black Pawn", /obj/structure/chess/blackpawn, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
		new /datum/stack_recipe("Black Rook", /obj/structure/chess/blackrook, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
		new /datum/stack_recipe("Black Knight", /obj/structure/chess/blackknight, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
		new /datum/stack_recipe("Black Bishop", /obj/structure/chess/blackbishop, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
		new /datum/stack_recipe("Black Queen", /obj/structure/chess/blackqueen, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
		new /datum/stack_recipe("Black King", /obj/structure/chess/blackking, 2, time = 1 SECONDS, one_per_turf = TRUE, on_floor = TRUE), \
```
Time and boolean defines

Rohesie commented on code/game/objects/structures/chess.dm:
```diff
@@ -0,0 +1,76 @@
+	desc = "This is weird. Please inform adminstration on how you managed to get the parent chess piece. Thanks!"
+	max_integrity = 100
+
+/obj/structure/chess/wrench_act(mob/user, obj/item/tool)
+	to_chat(user, "<span class='notice'>You take apart the chess piece.</span>")
+	var/obj/item/stack/sheet/metal/M = new (drop_location(), 2)
+	M.add_fingerprint(user)
+	tool.play_tool_sound(src)
+	qdel(src)
+	return TRUE
```
Add a small  `do_after()` to this?

KathrinBailey replied:
Thanks for the help. <3

Rohesie commented on code/game/objects/structures/beds_chairs/sofa.dm:
```diff
@@ -4,6 +4,28 @@
+	var/mutable_appearance/armrest
+
+/obj/structure/chair/sofa/Initialize()
+	armrest = mutable_appearance(icon, "[icon_state]_armrest", ABOVE_MOB_LAYER)
```
We can improve this code by using a globally cached visual object, which will have its own layer.
Poke me if you need help with that.

KathrinBailey committed:
Rad insulation changes on open() and close() instead

I did this code in 2019 and I was just trying to do anything to make it work.

KathrinBailey committed:
Rohesie's suggested change.

Co-authored-by: Rohesie <rohesie@gmail.com>

Rohesie commented on code/game/machinery/doors/shutters.dm:
```diff
@@ -15,3 +17,36 @@
+/obj/machinery/door/poddoor/shutters/radiation/preopen
+	icon_state = "open"
+	density = FALSE
+	opacity = 0
```
```suggestion
	opacity = FALSE
```

KathrinBailey committed:
Rohesie's suggested change.

Co-authored-by: Rohesie <rohesie@gmail.com>

optimumtact commented:
You shouldn't be in the habit of putting a bunch of unrelated changes into one PR

zxaber commented:
Maybe make the radiation doors use magenta instead of black?

![image](https://user-images.githubusercontent.com/37497534/93712109-5f92fc80-fb08-11ea-9358-204f60ab8a18.png)

Both color schemes are used (the black and yellow moreso internationally) but the magenta and yellow mix really help them pop out against other yellow and black signs.

Fikou commented:
can you change the chess pieces in holodeck to these new ones?
also the windowed shutters look really weird

MrDoomBringer commented:
I'm against the design philosophy behind windowed shutters. Extra protection that shutters give should come with the tradeoff that you don't know what's waiting for you on the other side.

LemonInTheDark commented:
I don't like having so much stuff in one pr, atomization allows us to pick and choose what we're interested in, and allows for faster merges on similar stuff. I could kind of understand if they were all related, but they're not. 

For instance:
I really like the couches, I'm a little worried about if the guy who made them is alright with us using them, but that's not a huge issue at all. They look great and what you've done makes a lot of sense, and feels like how it should have always been.

On another note, radiation shutters 
They do not work that way, insulation is a multiplier that effects the strength of the radiation wave, an insulation of 0.9 takes off 10% of a wave, 0 takes off 100%, this doesn't account for width of course, but in theory at least. In practice it would take off about 1- ((1-0.9)/ (9 tiles away on most sm setups)) or (1 - ~0.01) * intensity. I know that cits rad system and ours is not the same, but please look into the radiation.dm defines file, it has the defines for you to use. 

here's the math if you're interested. `intensity *= (1-((1-thing.rad_insulation)/width))`

Negative numbers are most certainly unexpected input, and may cause more radiation, not less, or perhaps even negative radiation in the right circumstances. Do not do this.

I don't really like the idea either, one of the sm's main threats is radiation, and I don't want to lessen it to any large degree.

Chess pieces are interesting, we already have a chess set in the holodeck, but it could do with new sprites in my codersprite fueled opinion.

As a whole, I'm very glad you're adding new things, and I really like a good deal of them, but others give me pause. I'd love to merge couches, but I can't do that without adding something I'm not interested in. That is the issue with this sort of position.

KathrinBailey committed:
Rad insulation + Do after proc

Thanks a ton to Rohesie for helping me make this comply with TG standards.

Tlaltecuhtli commented:
why shutter with windows?? they look bad and useless

KathrinBailey commented:
> I don't like having so much stuff in one pr, atomization allows us to pick and choose what we're interested in, and allows for faster merges on similar stuff. I could kind of understand if they were all related, but they're not.

Sorry, I was thinking at the time that it's all technically kinda furniture, but that's a bit too much of a stretch.

> For instance:
> I really like the couches, I'm a little worried about if the guy who made them is alright with us using them, but that's not a huge issue at all. They look great and what you've done makes a lot of sense, and feels like how it should have always been.

They've been on three different servers for over a year now I think. I did this when I couldn't sprite for heck. If I was to do this PR from scratch today I'd probably try making the sofa sprites myself (although I doubt I could surpass the quality of the sprites I'm taking from presumably Eris).

> On another note, radiation shutters
> They do not work that way, insulation is a multiplier that effects the strength of the radiation wave, an insulation of 0.9 takes off 10% of a wave, 0 takes off 100%, this doesn't account for width of course, but in theory at least. In practice it would take off about 1- ((1-0.9)/ (9 tiles away on most sm setups)) or (1 - ~0.01) * intensity. I know that cits rad system and ours is not the same, but please look into the radiation.dm defines file, it has the defines for you to use.
> 
> here's the math if you're interested. `intensity *= (1-((1-thing.rad_insulation)/width))`
> Negative numbers are most certainly unexpected input, and may cause more radiation, not less, or perhaps even negative radiation in the right circumstances. Do not do this.
> I don't really like the idea either, one of the sm's main threats is radiation, and I don't want to lessen it to any large degree.

Aaaah thanks a ton. I was under the impression that rad_insulation was supposed to affect how much I get irradiated, when in actuality, if I'm understanding you correctly, it's supposed to affect how many tiles the radiation spreads.
For now I put the rad insulation to 0.1 and amended the desc to reflect it. I'm happy to change the rad_insulation value to whatever you desire. You still get irradiated as much as before so the threat of radiation is still there.
(for context, last time I checked rwalls were 0.6, walls were 0.7 and open turfs of course 1.0)

> Chess pieces are interesting, we already have a chess set in the holodeck, but it could do with new sprites in my codersprite fueled opinion.

I could add these pieces into the holodeck, but the holodeck isn't wide enough, so it'd be pretty scuffed. Up to you if you'd prefer them there or not.

fludd12 commented:
> I could add these pieces into the holodeck, but the holodeck isn't wide enough, so it'd be pretty scuffed. Up to you if you'd prefer them there or not.

You could use a variant, like [minichess](https://en.wikipedia.org/wiki/Minichess). There's example rules for 4x, 5x, and 6x chess; I don't remember the exact dimensions of our holodeck, but one of those should be functional.

EOBGames commented:
> > I could add these pieces into the holodeck, but the holodeck isn't wide enough, so it'd be pretty scuffed. Up to you if you'd prefer them there or not.
> 
> You could use a variant, like [minichess](https://en.wikipedia.org/wiki/Minichess). There's example rules for 4x, 5x, and 6x chess; I don't remember the exact dimensions of our holodeck, but one of those should be functional.

the holodeck was recently expanded, and can now accommodate a full sized chess board IIRC

KathrinBailey commented:
> > > I could add these pieces into the holodeck, but the holodeck isn't wide enough, so it'd be pretty scuffed. Up to you if you'd prefer them there or not.
> > 
> > 
> > You could use a variant, like [minichess](https://en.wikipedia.org/wiki/Minichess). There's example rules for 4x, 5x, and 6x chess; I don't remember the exact dimensions of our holodeck, but one of those should be functional.
> 
> the holodeck was recently expanded, and can now accommodate a full sized chess board IIRC

I saw that on CC Z but I don't think the in-rotation maps have a larger holodeck to use them yet.

LemonInTheDark commented:
They do, see #52793 
Thanks for giving it a read, something like 0.5 would be good, I think there's a define for that you can use too.

KathrinBailey committed:
Adds chess pieces to the holodeck

The large holodeck area will now use actual chess pieces instead of cardboard cutouts renamed to be different chess pieces. This should make the game less confusing and awkward.

KathrinBailey committed:
Rad shutters use defines for rad_insulation

Also 0.5 insulation as requested by Lemons.

KathrinBailey commented:
@LemonInTheDark 
Rad insulation is now the define for 0.5 and the pieces are in the holodeck.

Rohesie reviewed (approved):
Code is fine. Pending art review.

Twaticus reviewed (approved):
the only real issue is the window shutter sprite lookin kinda lame but theres potential for them to look great after the rework
